CHANGELOG.md

```diff
@@ -2,6 +2,142 @@
 
 ## [Unreleased]
 
+## Changes
+[#352](https://github.com/sharksforarms/deku/pull/352) replaced the byte slice interface with new interfaces depending on `io::Read`.
+This brings massive performance and usability improvements.
+- `from_bytes(..)` was replaced with `from_reader(..)`:
+
+old:
+```rust
+#[derive(Debug, PartialEq, DekuRead, DekuWrite)]
+struct TestDeku(#[deku(bits = 4)] u8);
+
+let test_data: Vec<u8> = [0b0110_0110u8, 0b0101_1010u8].to_vec();
+
+let ((rest, i), ret_read) = TestDeku::from_bytes((&test_data, 0)).unwrap();
+assert_eq!(TestDeku(0b0110), ret_read);
+assert_eq!(2, rest.len());
+assert_eq!(4, i);
+```
+
+new:
+```rust
+#[derive(Debug, PartialEq, DekuRead, DekuWrite)]
+struct TestDeku(#[deku(bits = 4)] u8);
+
+let test_data: Vec<u8> = [0b0110_0110u8, 0b0101_1010u8].to_vec();
+let mut c = std::io::Cursor::new(test_data);
+
+let (amt_read, ret_read) = TestDeku::from_reader((&mut c, 0)).unwrap();
+assert_eq!(TestDeku(0b0110), ret_read);
+assert_eq!(amt_read, 4);
+```
+
+- The more internal (with context) `read(..)` was replaced with `from_reader_with_ctx(..)`.
+Now unused variables `deku::input`, `deku::input_bits`, and `deku::rest` were removed. `deku::reader` is the replacement.
```

Comment:
Pretty much.

deku::input was removed, as we don't cache the bytes read from `reader`. This could be added in, but I didn't want to add more to that hot loop.

I used that in one of my crates, and I just replaced it with a wrapper around the `read(..)`: https://github.com/rsadsb/adsb_deku/pull/212/files#diff-2149937cf9daa006df48008e1a47b5fac20f39821233df371e60e75c1abd4a66R158.

I'm not against adding a `deku::input` back in with the cache, as long as perf doesn't suffer like it did when I first tested it.

Reply:
Ok, would it be possible to provide an example in the examples folder and update the changelog to point to the alternative? This may avoid people creating an issue :sweat_smile: